8.x/theme-development/helpers.md

```diff
@@ -2,4 +2,2097 @@
+}
+```
+
+
+### get_product_similar_products()
+The get_product_similar_products function declares a similarity key for a product. For example, a car with the bmw brand is considered. This function returns products that are similar to this machine but have a different brand.
+
+```php
+if (!function_exists('get_product_similar_products')) {
+    /**
+     * @param Product $product
+     * @param int $count
+     * @param int $key_id
+     * @return Product[]
+     */
+    function get_product_similar_products(Product $product, int $count = 5, int $key_id = 0)
+    {
+        if ($key_id === 0) {
+            $key_id = $product->attributeKeys()->where("is_sortable", true)
+                ->pluck("p_structure_attr_keys.id")->get(0);
+        }
+
+        $product_key_values = $product->pAttributes()
+            ->where('p_structure_attr_key_id', $key_id)
+            ->pluck('p_structure_attr_value_id')->toArray();
+
+        $product_other_keys_attrs = $product->pAttributes()
+            ->where('p_structure_attr_key_id', '!=', $key_id)->get();
+        $product_other_keys_values = [];
+        foreach ($product_other_keys_attrs as $product_key_value) {
+            $item['key'] = $product_key_value->p_structure_attr_key_id;
+            $item['value'] = $product_key_value->p_structure_attr_value_id;
+            array_push($product_other_keys_values, $item);
+        }
+
+        $products = [];
+        if ($key_id !== 0) {
+            $product_structure = $product->productStructure;
+            if ($product_structure != null) {
+                $products = $product_structure->products()->where('directory_id',
+                    $product->directory_id)->mainModels()->visible()->except($product->id)
+                    ->whereHas('pAttributes', function ($q1) use ($key_id, $product_key_values) {
+                        $q1->where('p_structure_attr_key_id', $key_id)
+                            ->whereNotIn('p_structure_attr_value_id', $product_key_values);
+                    })->where(function ($q2) use ($key_id, $product_other_keys_values) {
+                        foreach ($product_other_keys_values as $item) {
+                            $key = $item['key'];
+                            $value = $item['value'];
+                            $q2->whereHas('pAttributes', function ($q3) use ($key_id, $key, $value) {
+                                $q3->where('p_structure_attr_key_id', '!=', $key_id);
+                                $q3->where([
+                                    'p_structure_attr_key_id' => $key,
+                                    'p_structure_attr_value_id' => $value
+                                ]);
+                            });
+                        }
+                    })->orderBy("priority", "ASC")->take($count)->get();
+            }
+        }
+        return $products;
+    }
+}
+```
+
+
+### get_related_products_with_directory_level()
+Returns similar products up to level 1 of a directory.
```

Comment:
Returns related products having the same parent in which they are in specific higher levels passed by `$level`.